subsys/bluetooth/services/dis.c

```diff
@@ -45,105 +48,265 @@ static struct dis_pnp dis_pnp_id = {
-#define BT_DIS_FW_REV_STR_REF		CONFIG_BT_DIS_FW_REV_STR
-#define BT_DIS_HW_REV_STR_REF		CONFIG_BT_DIS_HW_REV_STR
-#define BT_DIS_SW_REV_STR_REF		CONFIG_BT_DIS_SW_REV_STR
+#define BT_DIS_MODEL_REF             CONFIG_BT_DIS_MODEL
+#define BT_DIS_MANUF_REF             CONFIG_BT_DIS_MANUF
+#define BT_DIS_SERIAL_NUMBER_STR_REF CONFIG_BT_DIS_SERIAL_NUMBER_STR
+#define BT_DIS_FW_REV_STR_REF        CONFIG_BT_DIS_FW_REV_STR
+#define BT_DIS_HW_REV_STR_REF        CONFIG_BT_DIS_HW_REV_STR
+#define BT_DIS_SW_REV_STR_REF        CONFIG_BT_DIS_SW_REV_STR
+#define BT_DIS_UDI_LABEL_STR_REF     CONFIG_BT_DIS_UDI_LABEL_STR
+#define BT_DIS_UDI_DI_STR_REF        CONFIG_BT_DIS_UDI_DI_STR
+#define BT_DIS_UDI_ISSUER_STR_REF    CONFIG_BT_DIS_UDI_ISSUER_STR
+#define BT_DIS_UDI_AUTHORITY_STR_REF CONFIG_BT_DIS_UDI_AUTHORITY_STR
+#define BT_DIS_IEEE_RCDL_STR_REF     CONFIG_BT_DIS_IEEE_RCDL_STR
 
 #endif /* CONFIG_BT_DIS_SETTINGS */
 
-static ssize_t read_str(struct bt_conn *conn,
-			  const struct bt_gatt_attr *attr, void *buf,
-			  uint16_t len, uint16_t offset)
+static ssize_t read_str(struct bt_conn *conn, const struct bt_gatt_attr *attr, void *buf,
+			uint16_t len, uint16_t offset)
 {
 	return bt_gatt_attr_read(conn, attr, buf, len, offset, attr->user_data,
 				 strlen(attr->user_data));
 }
 
 #if CONFIG_BT_DIS_PNP
-static ssize_t read_pnp_id(struct bt_conn *conn,
-			   const struct bt_gatt_attr *attr, void *buf,
+static ssize_t read_pnp_id(struct bt_conn *conn, const struct bt_gatt_attr *attr, void *buf,
 			   uint16_t len, uint16_t offset)
 {
-	return bt_gatt_attr_read(conn, attr, buf, len, offset, &dis_pnp_id,
-				 sizeof(dis_pnp_id));
+	return bt_gatt_attr_read(conn, attr, buf, len, offset, &dis_pnp_id, sizeof(dis_pnp_id));
+}
+#endif
+
+#if CONFIG_BT_DIS_SYSTEM_ID
+static ssize_t read_system_id(struct bt_conn *conn, const struct bt_gatt_attr *attr, void *buf,
+			      uint16_t len, uint16_t offset)
+{
+	return bt_gatt_attr_read(conn, attr, buf, len, offset, dis_system_id,
+				 sizeof(dis_system_id));
+}
+#endif
+
+#if CONFIG_BT_DIS_UDI
+#define BT_DIS_UDI_FLAG_LABEL     (!!BT_DIS_UDI_LABEL_STR_REF[0])
+#define BT_DIS_UDI_FLAG_DI        (!!BT_DIS_UDI_DI_STR_REF[0])
+#define BT_DIS_UDI_FLAG_ISSUER    (!!BT_DIS_UDI_ISSUER_STR_REF[0])
+#define BT_DIS_UDI_FLAG_AUTHORITY (!!BT_DIS_UDI_AUTHORITY_STR_REF[0])
+#define BT_DIS_UDI_FLAGS                                                                           \
+	(BT_DIS_UDI_FLAG_LABEL | (BT_DIS_UDI_FLAG_DI << 1) | (BT_DIS_UDI_FLAG_ISSUER << 2) |       \
+	 (BT_DIS_UDI_FLAG_AUTHORITY << 3))
+
+/*
+ * UDI for medical devices contains a flag and 4 different null-terminated strings that may have
+ * unknown length. This requires its own encode method.
+ */
+static void read_udi_subval(const char *str, uint16_t str_len, char *buf, uint16_t *bytes_read,
+			    uint16_t *index, uint16_t len, uint16_t offset)
+{
+	/* String should not be with included null-terminator if empty */
+	if (str_len == 0) {
+		return;
+	}
```

Comment:
if `bytes_read == len` we can terminate early as well.

Similarly if `index + str_len < offset` we can also terminate early and just do `*index += str_len`